File: Python/TuringMachine.py
class TuringMachine:
    def __init__(self, input_string):
        # Initialize the tape with input and add extra space on both sides
        self.tape = list(" " * 100 + input_string + " " * 100)
        self.head = 100  # Start at the middle of the tape
        self.state = "q0"  # Initial state

    def execute(self):
        while self.state not in ["accept", "reject"]:
            current_symbol = self.tape[self.head]

            if self.state == "q0":
                if current_symbol == 'a':
                    self.tape[self.head] = 'X'
                    self.head += 1
                    self.state = "q1"
                elif current_symbol == ' ':
                    self.state = "accept"
                else:
                    self.state = "reject"

            elif self.state == "q1":
                if current_symbol in ['a', 'X']:
                    self.head += 1
                elif current_symbol == 'b':
                    self.tape[self.head] = 'Y'
                    self.head += 1
                    self.state = "q2"
                else:
                    self.state = "reject"

            elif self.state == "q2":
                if current_symbol in ['b', 'Y']:
                    self.head += 1
                elif current_symbol == 'c':
                    self.tape[self.head] = 'Z'
                    self.head -= 1
                    self.state = "q3"
                else:
                    self.state = "reject"

            elif self.state == "q3":
                if current_symbol in ['b', 'Y']:
                    self.head -= 1
                elif current_symbol == 'X':
                    self.head += 1
                    self.state = "q0"
                else:
                    self.state = "reject"

        if self.state == "accept":
            print("Input is accepted by the Turing Machine.")
        elif self.state == "reject":
            print("Input is rejected by the Turing Machine.")

    def print_tape(self):
        print(''.join(self.tape).strip())

if __name__ == "__main__":
    input_string = "aabbcc"
    tm = TuringMachine(input_string)
    tm.execute()
    tm.print_tape()

File: Java/TuringMachine.java
import java.util.Arrays;

class TuringMachine {
    private char[] tape;
    private int head;
    private String state;

    public TuringMachine(String input) {
        // Initialize the tape with the input and extra space on both sides
        tape = new char[300];
        Arrays.fill(tape, ' ');
        System.arraycopy(input.toCharArray(), 0, tape, 100, input.length());
        head = 100; // Start at the middle of the tape
        state = "q0"; // Initial state
    }

    public void execute() {
        while (!state.equals("accept") && !state.equals("reject")) {
            char currentSymbol = tape[head];
            switch (state) {
                case "q0":
                    if (currentSymbol == 'a') {
                        tape[head] = 'X';
                        head++;
                        state = "q1";
                    } else if (currentSymbol == ' ') {
                        state = "q_accept";
                    } else {
                        state = "q_reject";
                    }
                    break;

                case "q1":
                    if (currentSymbol == 'a' || currentSymbol == 'X') {
                        head++;
                    } else if (currentSymbol == 'b') {
                        tape[head] = 'Y';
                        head++;
                        state = "q2";
                    } else {
                        state = "q_reject";
                    }
                    break;

                case "q2":
                    if (currentSymbol == 'b' || currentSymbol == 'Y') {
                        head++;
                    } else if (currentSymbol == 'c') {
                        tape[head] = 'Z';
                        head--;
                        state = "q3";
                    } else {
                        state = "q_reject";
                    }
                    break;

                case "q3":
                    if (currentSymbol == 'b' || currentSymbol == 'Y') {
                        head--;
                    } else if (currentSymbol == 'X') {
                        head++;
                        state = "q0";
                    } else {
                        state = "q_reject";
                    }
                    break;

                case "q_accept":
                    System.out.println("Input is accepted by the Turing Machine.");
                    state = "accept";
                    break;

                case "q_reject":
                    System.out.println("Input is rejected by the Turing Machine.");
                    state = "reject";
                    break;

                default:
                    state = "q_reject";
            }
        }
    }

    public void printTape() {
        System.out.println(new String(tape).trim());
    }

    public static void main(String[] args) {
        String input = "aabbcc";
        TuringMachine tm = new TuringMachine(input);
        tm.execute();
        tm.printTape();
    }
}
